sklearn/naive_bayes.py

```diff
@@ -1164,6 +1180,21 @@ def _init_counters(self, n_effective_classes, n_features):
         self.category_count_ = [np.zeros((n_effective_classes, 0))
                                 for _ in range(n_features)]
 
+    def _init_n_categories(self, X):
+        feature_n_categories = X.max(axis=0) + 1
+        if self.min_categories is not None:
+            # unsafe casting allows float to be coerced to int
```

Comment:
Do we want this specified as a float?

Reply:
I do not think so. `n_categories_[i] - 1` will later be passed into `np.pad`, which will fail for float inputs.

We could handle float/integer condition later, but then accessing the `n_categories_` attribute gets weird, e.g.,
```
>>> clf = CategoricalNB(alpha=1, fit_prior=False, min_categories=[3., 4.2])
>>> clf.fit(X, y)
>>> clf.n_categories_
array([3., 4.2])
```
What does 4.2 here mean? It is kind of ambiguous, because it does not indicate how the value is rounded later on.

I think this does bring up the question of whether it is preferable to round up (`np.ceil`), as opposed to truncating (coercing to `int`).

Reply:
So I'm asking why we should support floats at all rather than raise an error

Reply:
Thanks for the additional clarification!

I think supporting it is desirable if the user wants to do some sort of calculation to determine the minimum number of categories. E.g., with heavy influence from `OrdinalEncoder` examples:

```
>>> from sklearn.preprocessing import OrdinalEncoder
>>> enc = OrdinalEncoder()
>>> X = [['Male', 1], ['Female', 3], ['Female', 2]]
>>> enc.fit(X)
OrdinalEncoder()
>>> enc.categories_
[array(['Female', 'Male'], dtype=object), array([1, 2, 3], dtype=object)]
>>> enc.transform([['Female', 3], ['Male', 1]])
array([[0., 2.],
       [1., 0.]])
>>> min_cat = enc.transform([['Female', 3], ['Male', 1]]).max(axis=0) + 1
>>> clf = CategoricalNB(min_categories=min_cat)
```

If we were to disallow floats, `min_cat` would need to be cast to an `int` first, either with something like
```
min_cat = min_cat.astype('int')
```

or if specific rounding behavior is desired
```
min_cat = np.ceil(min_cat).astype('int')
```

It seems like `CategoricalNB` should be able to handle some of this work for you.

So the trade-off is between seamlessly handling this behavior, and perhaps being ambitious in assuming the user's intent.

Do you have any thoughts/preferences one way or the other?

Reply:
I agree with @jnothman to not silently let pass float which will be cast to float. Things can go wrong and we will not catch this up.

Reply:
Okay, I have remove 'unsafe'.